packages/react-native-quick-crypto/cpp/ed25519/HybridEdKeyPair.cpp

```diff
@@ -0,0 +1,233 @@
+#include "HybridEdKeyPair.hpp"
+
+#include <memory>
+#include <string>
+
+namespace margelo::nitro::crypto {
+
+std::shared_ptr<Promise<void>>
+HybridEdKeyPair::generateKeyPair(
+  double publicFormat,
+  double publicType,
+  double privateFormat,
+  double privateType,
+  const std::optional<std::string>& cipher,
+  const std::optional<std::shared_ptr<ArrayBuffer>>& passphrase
+) {
+  // get owned NativeArrayBuffers before passing to sync function
+  std::optional<std::shared_ptr<ArrayBuffer>> nativePassphrase = std::nullopt;
+  if (passphrase.has_value()) {
+    nativePassphrase = ToNativeArrayBuffer(passphrase.value());
+  }
+
+  return Promise<void>::async(
+    [this, publicFormat, publicType, privateFormat, privateType, cipher,
+     nativePassphrase]() {
+      this->generateKeyPairSync(
```

Comment:
Just a random side-tip; technically `this` could already be deleted in here when this runs.
Imagine this JS code:

```ts
const promise = HybridEdKeyPair.generateKeyPair()
const actual = await promise
```

`HybridEdKeyPair` could technically be deleted by the JS VM before the Promise runs.
I am 100% sure that this will never happen though, because the JS VM doesn't delete objects that fast, and also people will always use `await` and the Promise never leaves the scope of the object.

So we're fine. But just to keep this in mind.
E.g. for long running functions/ticks we would need `shared()` (from `enable_shared_from_this()`) to get a safe ref-counted shared_ptr to `this`.

Reply:
This makes sense.  I'd love to see an example of your suggestion to do it "the right way".

When I add node.js compatibility to this really fast class implementation, we'll have static / self-contained fns where `this` will be available the entire time.

But I get what you are saying.

Reply:
You are already doing it the "right way" in this case. The approach with `shared()` would be a performance overhead which is unnecessary since it can never get deleted in this specific use-case.